cpp/src/arrow/compute/kernels/vector_replace_test.cc

```diff
@@ -580,6 +588,80 @@ TEST_F(TestReplaceDayTimeInterval, ReplaceWithMask) {
+               this->array("[]"));
+  this->Assert(ReplaceWithMask,
+               this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4]]"),
+               this->mask("[false, false, false, false]"), this->array("[]"),
+               this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4]]"));
+  this->Assert(ReplaceWithMask,
+               this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4]]"),
+               this->mask("[true, true, true, true]"),
+               this->array("[[3, 4, -2], [3, 4, -2], [3, 4, -2], [3, 4, -2]]"),
+               this->array("[[3, 4, -2], [3, 4, -2], [3, 4, -2], [3, 4, -2]]"));
+  this->Assert(ReplaceWithMask,
+               this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4]]"),
+               this->mask("[null, null, null, null]"), this->array("[]"),
+               this->array("[null, null, null, null]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], null]"),
+               this->mask("[false, false, false, false]"), this->array("[]"),
+               this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], null]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], null]"),
+               this->mask("[true, true, true, true]"),
+               this->array("[[3, 4, -2], [3, 4, -2], [3, 4, -2], [3, 4, -2]]"),
+               this->array("[[3, 4, -2], [3, 4, -2], [3, 4, -2], [3, 4, -2]]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], null]"),
+               this->mask("[null, null, null, null]"), this->array("[]"),
+               this->array("[null, null, null, null]"));
+  this->Assert(
+      ReplaceWithMask,
+      this->array("[[1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4], [1, 2, 4]]"),
+      this->mask("[true, true, false, false, null, null]"),
+      this->array("[[3, 4, -2], null]"),
+      this->array("[[3, 4, -2], null, [1, 2, 4], [1, 2, 4], null, null]"));
+  this->Assert(ReplaceWithMask, this->array("[null, null, null, null, null, null]"),
+               this->mask("[true, true, false, false, null, null]"),
+               this->array("[[3, 4, -2], null]"),
+               this->array("[[3, 4, -2], null, null, null, null, null]"));
+
+  this->Assert(ReplaceWithMask, this->array("[]"), this->mask("[]"),
+               this->scalar("[7, 0, 8]"), this->array("[]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2]]"),
+               this->mask("[true, true]"), this->scalar("[7, 0, 8]"),
+               this->array("[[7, 0, 8], [7, 0, 8]]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2]]"),
+               this->mask("[true, true]"), this->scalar("null"),
+               this->array("[null, null]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2], [-5, 6, 7]]"),
+               this->mask("[true, false, null]"), this->scalar("[7, 0, 8]"),
```

Comment:
Slightly unrelated to this PR, but it seems that amongst the `ReplaceWithMask` tests, the mask is always of the form (regex) `true* (false|null)*`, meaning that the replacement index is always trivially equal to the source index. Is that deliberate?

Reply:
It's not deliberate. I'll update the tests to avoid this.